core/src/main/scala/org/apache/spark/sql/SnappySession.scala

```diff
@@ -1775,6 +1783,44 @@ object SnappySession extends Logging {
 
       allLiterals.foreach(_.collectedForPlanCaching = true)
     }
+
+    logDebug(s"qe.executedPlan = ${df.queryExecution.executedPlan}")
+
+    // This part is the defensive coding for all those cases where Tokenization
+    // support is not smart enough to deal with cases where the execution plan
+    // is modified in such a way that we cannot track those constants which
+    // need to be replaced in the subsequent execution
+    if (key != null) {
+      val nocaching = session.getContextObject[Boolean](
+        CachedPlanHelperExec.NOCACHING_KEY).getOrElse(false)
+      if (nocaching /* || allallbroadcastplans.nonEmpty */) {
+        logDebug(s"Invalidating the key because explicit nocaching")
+        key.invalidatePlan()
+      }
+      else {
+        val params1 = getAllParamLiterals(executedPlan)
+        if (allLiterals.length != params1.length || !params1.sameElements(key.pls)) {
+          logDebug(s"Invalidating the key because nocaching " +
+              s"allLiterals.length = ${allLiterals.length}," +
+              s" params1.length = ${params1.length} and key.pls = ${key.pls.length}")
+          key.invalidatePlan()
+        }
+        else if (params1.length != 0 ) {
+          params1.foreach(p => {
+            if (!allLiterals.exists(_.position == p.pos)) {
+              logDebug(s"No plan caching for sql ${key.sqlText} as " +
+                  s"part execution of query has happened like in scalar subqueries" +
+                  s" or plans with broadcast hashjoins")
+              key.invalidatePlan()
+            }
+          })
+        }
+      }
+    }
+
+    if (key != null && key.valid) {
+      logDebug(s"Plan caching will be used for sql ${key.sqlText}")
```

Comment:
Right now using it to really see whether as per expectation plan caching will be used or not. May be we can remove later.